### PR TITLE
#3133 Redundant null check in generated code when adding parameters 

### PR DESCRIPTION
**Pull request by zyberzebra**
- verify redundant null checks in mappings are avoided
- targets mapper methods that involve update mappings and ensures that null checks are not applied redundantly if already covered

### CODE REVIEW

File: processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java
Comment (zyberzebra) [inline processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java:1]: I used the code formatter in IntelliJ: `etc/mapstruct.xml`

Comment (filiphr) [reply processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java:1]: When you are formatting it is best to only format the changed lines and also when committing it is best to only stage changes that are not formatting changes.

I'll clean up the not needed formatting changes now, that's not a problem

Comment (zyberzebra) [reply processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java:1]: Aaaah... makes sense! Thanks for clearing that up 😁

**Comment by zyberzebra** [comment]
Somthing must be configured incorrectly in my IDE... I'm not sure why though

File: processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java [hunk 1]
```diff
@@ -2096,7 +2096,16 @@ private boolean doesNotNeedPresenceCheckForSourceParameter(PropertyMapping mappi
...
+        return mapping.getAssignment().isSourceReferenceParameter()
+            || isCheckedByGuardClause( mapping );
+    }
+
+    private boolean isCheckedByGuardClause(PropertyMapping mapping) {
+        boolean isUpdateMethod = getReturnTypeToConstruct() == null && isMapNullToDefault();
+        return !isUpdateMethod
+            && presenceChecksByParameter.size() == 1
+            && presenceChecksByParameter.containsKey( mapping.getSourceBeanName() )
+            && mapping.getAssignment().getSourcePresenceCheckerReference() == null;
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java:2108]: Why is this needed? All tests pass without this, did I remove some edge case in the existing test cases?

**Review by filiphr — COMMENTED** [review]
Nice work @zyberzebra. I've done some small polishing (see my commits). Let me know if I've removed too much.

Can you please also rebase this on top of main?

File: processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java (hunk 1 above)
Comment (zyberzebra) [reply processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java:2108]: Are you refering to only the last condition? 
`mapping.getAssignment().getSourcePresenceCheckerReference() == null`

Comment (filiphr) [reply processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java:2108]: Yes indeed, only the last condition

Comment (zyberzebra) [reply processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java:2108]: That is indeed a really good question.
I'm going to rebase and try a little if I can find out why I thought I need it.
If I can reproduce it, I'm going to remove the condition and update the PR.

Comment (filiphr) [reply processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java:2108]: Please also check if the class and methods I removed from the tests are causing everything to be green

Comment (zyberzebra) [reply processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java]: I checked out the older revision and none of my Testcases covered that line. 
Sadly I can't tell you my initial thought when I wrote that line.

**Comment by zyberzebra** [comment]
@filiphr Is just a rebase needed again?

**Comment by filiphr** [comment]
Not sure what the problem is @zyberzebra, the logs from the builds are gone.
